File: autocontext/src/autocontext/harness/core/events.py
"""Domain-agnostic event stream emitter with thread safety."""

from __future__ import annotations

import json
import logging
import threading
from collections.abc import Callable
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

logger = logging.getLogger(__name__)

EventCallback = Callable[[str, dict[str, Any]], None]


class EventStreamEmitter:
    def __init__(self, path: Path) -> None:
        self.path = path
        self._sequence = 0
        self._subscribers: list[EventCallback] = []
        self._lock = threading.Lock()

    def subscribe(self, callback: EventCallback) -> None:
        with self._lock:
            self._subscribers.append(callback)

    def unsubscribe(self, callback: EventCallback) -> None:
        with self._lock:
            self._subscribers.remove(callback)

    def emit(self, event: str, payload: dict[str, Any], channel: str = "generation") -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        with self._lock:
            self._sequence += 1
            seq = self._sequence
            subscribers = list(self._subscribers)
        line = {
            "ts": datetime.now(UTC).isoformat(),
            "v": 1,
            "seq": seq,
            "channel": channel,
            "event": event,
            "payload": payload,
        }
        with self.path.open("a", encoding="utf-8") as handle:
            handle.write(json.dumps(line, sort_keys=True) + "\n")
        for cb in subscribers:
            try:
                cb(event, payload)
            except Exception:
                try:
                    logger.debug("harness.core.events: suppressed Exception", exc_info=True)
                except Exception:
                    pass

File: ts/src/loop/events.ts
/**
 * Event stream emitter — NDJSON file + subscriber dispatch (AC-342).
 * Mirrors Python's autocontext/harness/core/events.py.
 */

import { appendFileSync, mkdirSync } from "node:fs";
import { dirname } from "node:path";

export type EventCallback = (event: string, payload: Record<string, unknown>) => void;

export class EventStreamEmitter {
  readonly path: string;
  #sequence = 0;
  #subscribers: EventCallback[] = [];

  constructor(path: string) {
    this.path = path;
  }

  subscribe(callback: EventCallback): void {
    this.#subscribers.push(callback);
  }

  unsubscribe(callback: EventCallback): void {
    const idx = this.#subscribers.indexOf(callback);
    if (idx !== -1) {
      this.#subscribers.splice(idx, 1);
    }
  }

  emit(
    event: string,
    payload: Record<string, unknown>,
    channel = "generation",
  ): void {
    // Ensure parent directory exists
    mkdirSync(dirname(this.path), { recursive: true });

    this.#sequence += 1;
    const seq = this.#sequence;
    const subscribersCopy = [...this.#subscribers];

    const line = {
      channel,
      event,
      payload,
      seq,
      ts: new Date().toISOString(),
      v: 1,
    };

    appendFileSync(this.path, JSON.stringify(line) + "\n", "utf-8");

    for (const cb of subscribersCopy) {
      try {
        cb(event, payload);
      } catch {
        // subscriber errors must never crash the loop
      }
    }
  }
}
